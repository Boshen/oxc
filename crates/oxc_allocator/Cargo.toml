[package]
name = "oxc_allocator"
authors.workspace = true
description.workspace = true
edition.workspace = true
homepage.workspace = true
keywords.workspace = true
license.workspace = true
repository.workspace = true
version.workspace = true

[dependencies]
bumpalo = { workspace = true }
blink-alloc = { workspace = true }
allocator-api2 = { workspace = true }
serde = { workspace = true }
