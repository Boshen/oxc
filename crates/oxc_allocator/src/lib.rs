use std::{
    alloc::Layout,
    fmt,
    hash::{Hash, Hasher},
    ops::{self, Deref},
    ptr::{self, NonNull},
};

use allocator_api2::alloc::AllocError;
use blink_alloc::BlinkAlloc;
// use bumpalo::Bump;
use serde::ser::{Serialize, Serializer};

// #[derive(Debug)]
pub struct Allocator {
    blink: BlinkAlloc,
}

impl Allocator {
    #[inline(always)]
    pub fn alloc<T>(&self, value: T) -> &mut T {
        let ptr = self.blink.allocate(Layout::new::<T>()).unwrap();
        let ptr = ptr.as_ptr().cast::<T>();
        unsafe {
            ptr.write(value);
        }
        unsafe { &mut *ptr }
    }

    #[inline(always)]
    pub fn alloc_str(&self, value: &str) -> &mut str {
        let ptr = self.blink.allocate(Layout::for_value(value)).unwrap();
        let ptr = ptr.as_ptr().cast::<u8>();
        let slice = unsafe { std::slice::from_raw_parts_mut(ptr, value.len()) };
        slice.copy_from_slice(value.as_bytes());
        unsafe { std::str::from_utf8_unchecked_mut(slice) }
    }
}

// SAFETY: Make Bump Sync and Send, it's our responsibility to never
// simultaneously mutate across threads.
unsafe impl Send for Allocator {}
unsafe impl Sync for Allocator {}

unsafe impl allocator_api2::alloc::Allocator for Allocator {
    #[inline(always)]
    fn allocate(&self, layout: Layout) -> Result<NonNull<[u8]>, AllocError> {
        // self.blink
        //     .try_alloc_layout(layout)
        //     .map(|ptr| unsafe {
        //         NonNull::new_unchecked(core::ptr::slice_from_raw_parts_mut(
        //             ptr.as_ptr(),
        //             layout.size(),
        //         ))
        //     })
        //     .map_err(|_| AllocError)

        self.blink.allocate(layout)
    }

    #[inline(always)]
    unsafe fn deallocate(&self, _ptr: NonNull<u8>, _layout: Layout) {}
}

// pub type Box<'a, T> = allocator_api2::boxed::Box<T, &'a Allocator>;
pub type Vec<'a, T> = allocator_api2::vec::Vec<T, &'a Allocator>;

impl Default for Allocator {
    fn default() -> Self {
        Self { blink: BlinkAlloc::new() }
    }
}

impl Deref for Allocator {
    type Target = BlinkAlloc;

    #[inline(always)]
    fn deref(&self) -> &Self::Target {
        &self.blink
    }
}

pub struct String<'a> {
    vec: Vec<'a, u8>,
}

impl<'a> String<'a> {
    #[inline(always)]
    pub fn new_in(alloc: &'a Allocator) -> Self {
        String { vec: Vec::new_in(alloc) }
    }

    #[inline(always)]
    pub fn as_str(&self) -> &str {
        unsafe { std::str::from_utf8_unchecked(&self.vec) }
    }

    #[inline(always)]
    pub fn push(&mut self, ch: char) {
        match ch.len_utf8() {
            1 => self.vec.push(ch as u8),
            _ => self.vec.extend_from_slice(ch.encode_utf8(&mut [0; 4]).as_bytes()),
        }
    }

    #[inline(always)]
    pub fn leak(self) -> &'a str {
        unsafe { std::str::from_utf8_unchecked(self.vec.leak()) }
    }

    #[inline(always)]
    pub fn from_str_in(s: &str, alloc: &'a Allocator) -> Self {
        let mut vec = Vec::with_capacity_in(s.len(), alloc);
        vec.extend_from_slice(s.as_bytes());
        String { vec }
    }

    #[inline(always)]
    pub fn push_str(&mut self, s: &str) {
        self.vec.extend_from_slice(s.as_bytes());
    }
}

impl<'a> fmt::Debug for String<'a> {
    #[inline(always)]
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self.as_str(), f)
    }
}

/// Bumpalo Box
pub struct Box<'alloc, T: ?Sized>(pub &'alloc mut T);

impl<'alloc, T> Box<'alloc, T> {
    #[must_use]
    pub fn into_inner(b: Self) -> T {
        // This pointer read is safe because the reference `self.0` is
        // guaranteed to be unique--not just now, but we're guaranteed it's not
        // borrowed from some other reference. This in turn is because we never
        // construct an alloc::Box with a borrowed reference, only with a fresh
        // one just allocated from a Bump.
        unsafe { ptr::read(b.0 as *mut T) }
    }
}

impl<'alloc, T: ?Sized> ops::Deref for Box<'alloc, T> {
    type Target = T;

    #[inline(always)]
    fn deref(&self) -> &T {
        self.0
    }
}

impl<'alloc, T: ?Sized> ops::DerefMut for Box<'alloc, T> {
    #[inline(always)]
    fn deref_mut(&mut self) -> &mut T {
        self.0
    }
}

impl<'alloc, T: ?Sized + fmt::Debug> fmt::Debug for Box<'alloc, T> {
    #[inline(always)]
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        self.0.fmt(f)
    }
}

impl<'alloc, T> PartialEq for Box<'alloc, T>
where
    T: PartialEq<T> + ?Sized,
{
    #[inline(always)]
    fn eq(&self, other: &Box<'alloc, T>) -> bool {
        PartialEq::eq(&**self, &**other)
    }
}

impl<'alloc, T> Serialize for Box<'alloc, T>
where
    T: Serialize,
{
    #[inline(always)]
    fn serialize<S>(&self, s: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        self.0.serialize(s)
    }
}

impl<'alloc, T: Hash> Hash for Box<'alloc, T> {
    #[inline(always)]
    fn hash<H: Hasher>(&self, state: &mut H) {
        self.0.hash(state);
    }
}
