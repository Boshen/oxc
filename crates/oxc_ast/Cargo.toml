[package]
name = "oxc_ast"
authors.workspace = true
description.workspace = true
edition.workspace = true
homepage.workspace = true
keywords.workspace = true
license.workspace = true
repository.workspace = true
version.workspace = true

[dependencies]
thiserror = { workspace = true }
compact_str = { workspace = true, features = ["serde"] }
