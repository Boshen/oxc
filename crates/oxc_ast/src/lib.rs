//! AST
//! NOTE: This is not compatible with estree.

#![feature(let_chains)]
#![feature(is_some_and)]

mod serialize;

pub mod ast;
pub mod ast_builder;
pub mod context;
pub mod node;
pub mod source_type;

pub use num_bigint::BigUint;

pub use self::ast::*;
pub use self::ast_builder::*;
pub use self::node::*;
pub use self::source_type::*;

pub type Atom = compact_str::CompactString;

// After experimenting with two types of boxed enum variants:
//   1.
//   ```
//      enum Expression {
//          Variant(Box<Struct>)
//      }
//      struct Struct {
//          expression: Expression
//      }
//   ```
//   2.
//   ```
//      enum Expression {
//          Variant(Struct)
//      }
//      struct Struct {
//          expression: Box<Expression>
//      }
//   ```
//  I have concluded that the first options is more performant and more ergonomic to use.
//  The following test make sure all enum variants are boxed, resulting 16 bytes for each enum.
//  Read `https://nnethercote.github.io/perf-book/type-sizes.html` for more details.
#[cfg(all(target_arch = "x86_64", target_pointer_width = "64"))]
#[test]
fn no_bloat_enum_sizes() {
    use std::mem::size_of;
    assert_eq!(size_of::<Statement>(), 16);
    assert_eq!(size_of::<Expression>(), 16);
    assert_eq!(size_of::<Declaration>(), 16);
    assert_eq!(size_of::<BindingPatternKind>(), 16);
    assert_eq!(size_of::<ModuleDeclarationKind>(), 16);
    assert_eq!(size_of::<ClassElement>(), 16);
    assert_eq!(size_of::<ExportDefaultDeclarationKind>(), 16);
    assert_eq!(size_of::<TSLiteral>(), 16);
    assert_eq!(size_of::<TSType>(), 16);
}
