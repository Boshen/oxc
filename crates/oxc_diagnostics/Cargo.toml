[package]
name = "oxc_diagnostics"
authors.workspace = true
description.workspace = true
edition.workspace = true
homepage.workspace = true
keywords.workspace = true
license.workspace = true
repository.workspace = true
version.workspace = true

[dependencies]
thiserror = { workspace = true }
miette = { workspace = true }
