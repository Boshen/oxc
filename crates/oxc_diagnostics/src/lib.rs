//! Diagnostics Wrapper
//! Exports `thiserror` and `miette`

use std::{cell::RefCell, ops::Deref, rc::Rc};

pub use miette;
pub use thiserror;

pub type Error = miette::Error;
pub type Severity = miette::Severity;
pub type Report = miette::Report;

pub type Result<T> = std::result::Result<T, Error>;

#[derive(Debug, Default, Clone)]
pub struct Diagnostics(Rc<RefCell<Vec<Error>>>);

impl Deref for Diagnostics {
    type Target = Rc<RefCell<Vec<Error>>>;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl Diagnostics {
    /// # Panics
    #[must_use]
    pub fn into_inner(self) -> Vec<Error> {
        Rc::try_unwrap(self.0).unwrap().into_inner()
    }
}
