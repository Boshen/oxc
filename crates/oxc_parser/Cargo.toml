[package]
name = "oxc_parser"
authors.workspace = true
description.workspace = true
edition.workspace = true
homepage.workspace = true
keywords.workspace = true
license.workspace = true
repository.workspace = true
version.workspace = true

[dependencies]
oxc_allocator = { path = "../oxc_allocator" }
oxc_ast = { path = "../oxc_ast" }
oxc_diagnostics = { path = "../oxc_diagnostics" }

bitflags = { workspace = true }
rustc-hash = { workspace = true }

phf = { version = "0.11.1", features = ["macros"] }
unicode-id-start = "1.0.3"
num-bigint = "0.4.3"

[dev-dependencies]
miette = { workspace = true, features = ["fancy-no-backtrace"] }
oxc_ast = { path = "../oxc_ast", features = ["serde_json"] }
serde_json = { workspace = true }
