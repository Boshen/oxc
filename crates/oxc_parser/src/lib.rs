//! Recursive Descent Parser for ECMAScript and TypeScript

#![allow(clippy::wildcard_imports)] // allow for use `oxc_ast::ast::*`

mod cursor;
mod list;
mod state;

mod js;
mod jsx;
mod ts;

mod lexer;

use oxc_allocator::Allocator;
use oxc_ast::{ast::Program, context::Context, AstBuilder, Node, SourceType};
use oxc_diagnostics::{Diagnostic, Diagnostics, Result};

use crate::{
    lexer::{Kind, Lexer, Token},
    state::ParserState,
};

#[derive(Debug)]
pub struct ParserReturn<'a> {
    pub program: Program<'a>,
    pub errors: Vec<Diagnostic>,
}

pub struct Parser<'a> {
    lexer: Lexer<'a>,

    /// SourceType: JavaScript or TypeScript, Script or Module, jsx support?
    source_type: SourceType,

    /// Source Code
    source: &'a str,

    /// All syntax errors from parser and lexer
    /// Note: favor adding to `Diagnostics` instead of raising Err
    errors: Diagnostics,

    /// The current parsing token
    token: Token,

    /// The end range of the previous token
    prev_token_end: usize,

    /// Parser state
    state: ParserState<'a>,

    /// Parsing context saved into every AST node
    ctx: Context,

    /// Ast builder for creating AST nodes
    ast: AstBuilder<'a>,
}

impl<'a> Parser<'a> {
    #[must_use]
    pub fn new(allocator: &'a Allocator, source: &'a str, source_type: SourceType) -> Self {
        let errors = Diagnostics::default();
        Self {
            lexer: Lexer::new(allocator, source, errors.clone(), source_type),
            source_type,
            source,
            errors,
            token: Token::default(),
            prev_token_end: 0,
            state: ParserState::default(),
            ctx: source_type.default_context(),
            ast: AstBuilder::new(allocator),
        }
    }

    #[must_use]
    pub fn allow_return_outside_function(mut self, allow: bool) -> Self {
        self.ctx = self.ctx.and_return(allow);
        self
    }

    /// Parser main entry point
    /// Returns an empty `Program` on unrecoverable error,
    /// Recoverable errors are stored inside `errors`.
    #[must_use]
    pub fn parse(mut self) -> ParserReturn<'a> {
        let program = match self.parse_program() {
            Ok(program) => program,
            Err(error) => {
                self.error(self.flow_error().unwrap_or(error));
                let program = self.ast.program(
                    Node::default(),
                    self.ast.new_vec(),
                    self.ast.new_vec(),
                    self.source_type,
                );
                program
            }
        };
        ParserReturn { program, errors: self.errors.borrow().clone() }
    }

    fn parse_program(&mut self) -> Result<Program<'a>> {
        // initialize cur_token and prev_token by moving onto the first token
        self.bump_any();

        let (directives, statements) =
            self.parse_directives_and_statements(/* is_top_level */ true)?;

        let node = Node::new(0, self.source.len(), self.ctx);
        Ok(self.ast.program(node, directives, statements, self.source_type))
    }

    /// Check for Flow declaration if the file cannot be parsed.
    /// The declaration must be [on the first line before any code](https://flow.org/en/docs/usage/#toc-prepare-your-code-for-flow)
    fn flow_error(&self) -> Option<Diagnostic> {
        if self.source_type.is_javascript()
            && (self.source.starts_with("// @flow") || self.source.starts_with("/* @flow */"))
        {
            return Some(Diagnostic::Flow(0..8));
        }
        None
    }

    /// Return error info at current token
    /// # Panics
    ///   * The lexer did not push a diagnostic when `Kind::Undetermined` is returned
    fn unexpected<T>(&self) -> Result<T> {
        // The lexer should have reported a more meaningful diagnostic
        // when it is a undetermined kind.
        if self.cur_kind() == Kind::Undetermined {
            return Err(self.errors.borrow_mut().pop().unwrap());
        }
        Err(Diagnostic::UnexpectedToken(self.current_range()))
    }

    /// Push a Syntax Error
    fn error(&mut self, error: Diagnostic) {
        self.errors.borrow_mut().push(error);
    }

    #[must_use]
    const fn ts_enabled(&self) -> bool {
        self.source_type.is_typescript()
    }
}

#[cfg(test)]
mod test {

    use super::*;

    #[test]
    fn smoke_test() {
        let allocator = Allocator::default();
        let source_type = SourceType::default();
        let source = "";
        let ret = Parser::new(&allocator, source, source_type).parse();
        assert!(ret.program.is_empty());
        assert!(ret.errors.is_empty());
    }

    #[test]
    fn flow_error() {
        let allocator = Allocator::default();
        let source_type = SourceType::default();
        let source = "// @flow\nasdf adsf";
        let ret = Parser::new(&allocator, source, source_type).parse();
        assert!(ret.program.is_empty());
        assert_eq!(ret.errors.first().unwrap().to_string(), "Flow is not supported");

        let source = "/* @flow */\n asdf asdf";
        let ret = Parser::new(&allocator, source, source_type).parse();
        assert!(ret.program.is_empty());
        assert_eq!(ret.errors.first().unwrap().to_string(), "Flow is not supported");
    }
}
