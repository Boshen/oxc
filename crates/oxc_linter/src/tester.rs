use std::path::PathBuf;

use oxc_allocator::Allocator;
use oxc_ast::SourceType;
use oxc_diagnostics::miette::{GraphicalReportHandler, GraphicalTheme, NamedSource};
use oxc_parser::Parser;
use oxc_semantic::SemanticBuilder;

use crate::{rules::RULES, Linter};

pub struct Tester {
    rule_name: &'static str,
    expect_pass: Vec<String>,
    expect_fail: Vec<String>,
    snapshot: String,
}

impl Tester {
    #[allow(clippy::needless_pass_by_value)]
    pub fn new<S: Into<String>>(
        rule_name: &'static str,
        expect_pass: Vec<S>,
        expect_fail: Vec<S>,
    ) -> Self {
        let expect_pass = expect_pass.into_iter().map(Into::into).collect::<Vec<_>>();
        let expect_fail = expect_fail.into_iter().map(Into::into).collect::<Vec<_>>();
        Self { rule_name, expect_pass, expect_fail, snapshot: String::new() }
    }

    pub fn test_and_snapshot(&mut self) {
        self.test_pass();
        self.test_fail();
        self.snapshot();
    }

    fn test_pass(&mut self) {
        for test in self.expect_pass.clone() {
            let passed = self.run(&test);
            assert!(passed, "expect test to pass: {test} {}", self.snapshot);
        }
    }

    fn test_fail(&mut self) {
        for test in self.expect_fail.clone() {
            let passed = self.run(&test);
            assert!(!passed, "expect test to fail: {test}");
        }
    }

    fn snapshot(&self) {
        let name = self.rule_name.replace('-', "_");
        insta::with_settings!({ prepend_module_to_snapshot => false, }, {
            insta::assert_snapshot!(name.clone(), self.snapshot, &name);
        });
    }

    fn run(&mut self, source_text: &str) -> bool {
        let name = self.rule_name.replace('-', "_");
        let allocator = Allocator::default();
        let path = PathBuf::from(name).with_extension("tsx");
        let source_type = SourceType::from_path(&path).expect("incorrect {path:?}");
        let ret = Parser::new(&allocator, source_text, source_type).parse();
        assert!(ret.errors.is_empty(), "{:?}", &ret.errors);
        let program = allocator.alloc(ret.program);
        let semantic = SemanticBuilder::new().build(program, ret.trivias);
        let semantic = std::rc::Rc::new(semantic);
        let rule = RULES.iter().find(|rule| rule.name() == self.rule_name).unwrap();
        let diagnostics = Linter::from_rules(vec![rule.clone()]).run(&semantic);
        if diagnostics.is_empty() {
            return true;
        }
        let handler = GraphicalReportHandler::new_themed(GraphicalTheme::unicode_nocolor());
        for diagnostic in diagnostics {
            let diagnostic = diagnostic.with_source_code(source_text.to_string());
            let diagnostic = diagnostic.with_source_code(NamedSource::new(
                path.to_string_lossy(),
                source_text.to_string(),
            ));
            handler.render_report(&mut self.snapshot, diagnostic.as_ref()).unwrap();
        }
        false
    }
}
