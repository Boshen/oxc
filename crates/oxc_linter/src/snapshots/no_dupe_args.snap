---
source: crates/oxc_linter/src/tester.rs
assertion_line: 53
expression: no_dupe_args
---

  × Identifier `b` has already been declared
   ╭─[no_dupe_args.tsx:1:1]
 1 │ function a(a, b, b) {}
   ·               ┬  ┬
   ·               │  ╰── It can not be redeclared here
   ·               ╰── `b` has already been declared here
   ╰────

  ⚠ eslint(no-dupe-args): Disallow duplicate arguments in function definitions
   ╭─[no_dupe_args.tsx:1:1]
 1 │ function a(a, b, b) {}
   ·               ─  ─
   ╰────
  help: Consider removing the duplicated argument

  × Identifier `a` has already been declared
   ╭─[no_dupe_args.tsx:1:1]
 1 │ function a(a, a, a) {}
   ·            ┬  ┬
   ·            │  ╰── It can not be redeclared here
   ·            ╰── `a` has already been declared here
   ╰────

  × Identifier `a` has already been declared
   ╭─[no_dupe_args.tsx:1:1]
 1 │ function a(a, a, a) {}
   ·               ┬  ┬
   ·               │  ╰── It can not be redeclared here
   ·               ╰── `a` has already been declared here
   ╰────

  ⚠ eslint(no-dupe-args): Disallow duplicate arguments in function definitions
   ╭─[no_dupe_args.tsx:1:1]
 1 │ function a(a, a, a) {}
   ·            ─  ─
   ╰────
  help: Consider removing the duplicated argument

  ⚠ eslint(no-dupe-args): Disallow duplicate arguments in function definitions
   ╭─[no_dupe_args.tsx:1:1]
 1 │ function a(a, a, a) {}
   ·               ─  ─
   ╰────
  help: Consider removing the duplicated argument

  × Identifier `a` has already been declared
   ╭─[no_dupe_args.tsx:1:1]
 1 │ function a(a, b, a) {}
   ·            ┬     ┬
   ·            │     ╰── It can not be redeclared here
   ·            ╰── `a` has already been declared here
   ╰────

  ⚠ eslint(no-dupe-args): Disallow duplicate arguments in function definitions
   ╭─[no_dupe_args.tsx:1:1]
 1 │ function a(a, b, a) {}
   ·            ─     ─
   ╰────
  help: Consider removing the duplicated argument

  × Identifier `a` has already been declared
   ╭─[no_dupe_args.tsx:1:1]
 1 │ function a(a, b, a, b) {}
   ·            ┬     ┬
   ·            │     ╰── It can not be redeclared here
   ·            ╰── `a` has already been declared here
   ╰────

  × Identifier `b` has already been declared
   ╭─[no_dupe_args.tsx:1:1]
 1 │ function a(a, b, a, b) {}
   ·               ┬     ┬
   ·               │     ╰── It can not be redeclared here
   ·               ╰── `b` has already been declared here
   ╰────

  ⚠ eslint(no-dupe-args): Disallow duplicate arguments in function definitions
   ╭─[no_dupe_args.tsx:1:1]
 1 │ function a(a, b, a, b) {}
   ·            ─     ─
   ╰────
  help: Consider removing the duplicated argument

  ⚠ eslint(no-dupe-args): Disallow duplicate arguments in function definitions
   ╭─[no_dupe_args.tsx:1:1]
 1 │ function a(a, b, a, b) {}
   ·               ─     ─
   ╰────
  help: Consider removing the duplicated argument

  × Identifier `b` has already been declared
   ╭─[no_dupe_args.tsx:1:1]
 1 │ var a = function(a, b, b) {}
   ·                     ┬  ┬
   ·                     │  ╰── It can not be redeclared here
   ·                     ╰── `b` has already been declared here
   ╰────

  ⚠ eslint(no-dupe-args): Disallow duplicate arguments in function definitions
   ╭─[no_dupe_args.tsx:1:1]
 1 │ var a = function(a, b, b) {}
   ·                     ─  ─
   ╰────
  help: Consider removing the duplicated argument

  × Identifier `a` has already been declared
   ╭─[no_dupe_args.tsx:1:1]
 1 │ var a = function(a, a, a) {}
   ·                  ┬  ┬
   ·                  │  ╰── It can not be redeclared here
   ·                  ╰── `a` has already been declared here
   ╰────

  × Identifier `a` has already been declared
   ╭─[no_dupe_args.tsx:1:1]
 1 │ var a = function(a, a, a) {}
   ·                     ┬  ┬
   ·                     │  ╰── It can not be redeclared here
   ·                     ╰── `a` has already been declared here
   ╰────

  ⚠ eslint(no-dupe-args): Disallow duplicate arguments in function definitions
   ╭─[no_dupe_args.tsx:1:1]
 1 │ var a = function(a, a, a) {}
   ·                  ─  ─
   ╰────
  help: Consider removing the duplicated argument

  ⚠ eslint(no-dupe-args): Disallow duplicate arguments in function definitions
   ╭─[no_dupe_args.tsx:1:1]
 1 │ var a = function(a, a, a) {}
   ·                     ─  ─
   ╰────
  help: Consider removing the duplicated argument

  × Identifier `a` has already been declared
   ╭─[no_dupe_args.tsx:1:1]
 1 │ var a = function(a, b, a) {}
   ·                  ┬     ┬
   ·                  │     ╰── It can not be redeclared here
   ·                  ╰── `a` has already been declared here
   ╰────

  ⚠ eslint(no-dupe-args): Disallow duplicate arguments in function definitions
   ╭─[no_dupe_args.tsx:1:1]
 1 │ var a = function(a, b, a) {}
   ·                  ─     ─
   ╰────
  help: Consider removing the duplicated argument

  × Identifier `a` has already been declared
   ╭─[no_dupe_args.tsx:1:1]
 1 │ var a = function(a, b, a, b) {}
   ·                  ┬     ┬
   ·                  │     ╰── It can not be redeclared here
   ·                  ╰── `a` has already been declared here
   ╰────

  × Identifier `b` has already been declared
   ╭─[no_dupe_args.tsx:1:1]
 1 │ var a = function(a, b, a, b) {}
   ·                     ┬     ┬
   ·                     │     ╰── It can not be redeclared here
   ·                     ╰── `b` has already been declared here
   ╰────

  ⚠ eslint(no-dupe-args): Disallow duplicate arguments in function definitions
   ╭─[no_dupe_args.tsx:1:1]
 1 │ var a = function(a, b, a, b) {}
   ·                  ─     ─
   ╰────
  help: Consider removing the duplicated argument

  ⚠ eslint(no-dupe-args): Disallow duplicate arguments in function definitions
   ╭─[no_dupe_args.tsx:1:1]
 1 │ var a = function(a, b, a, b) {}
   ·                     ─     ─
   ╰────
  help: Consider removing the duplicated argument

